src/advanced_inv.cpp

```diff
@@ -1892,27 +1997,34 @@ bool advanced_inventory::query_charges( aim_location destarea, const advanced_in
             amount = std::min( weightmax, amount );
         }
     }
+    // handle how many of armour type we can equip (max of 2 per type)
+    if(destarea == AIM_WORN) {
+        const auto &id = sitem.items.front()->typeId();
+        // how many slots are available for the item?
+        const int slots_available = input_amount - g->u.amount_worn(id);
+        // if something is going to the worn pane, there has to be at least 1
+        amount = std::min(slots_available, 1);
+    }
 
     // Now we have the final amount. Query if needed (either requested, or when
     // the destination can not hold all items).
-    if( askamount || amount < input_amount ) {
-        // moving several items (not charges!) from ground it currently not implemented.
-        // TODO: implement this properly, see the code where this is called from.
-        if( !by_charges && sitem.area != AIM_INVENTORY && sitem.area != AIM_WORN ) {
-            amount = 1;
-            return true;
-        }
-        std::string popupmsg = _( "How many do you want to move? (0 to cancel)" );
+    if( action == "MOVE_VARIABLE_ITEM" || amount < input_amount ) {
+        const int count = (by_charges) ? it.charges : sitem.stacks;
         // At this point amount contains the maximal amount that the destination can hold.
-        if( amount < input_amount ) {
-            popupmsg = string_format( _( "Destination can only hold %d! Move how many? (0 to cancel) " ),
-                                      amount );
-        }
-        // instead of prompting to move all, prompt to move all _but_ one
+        const char *msg = (amount < input_amount) ?
```

Comment:
Nitpicking: this (two strings with different format specifiers stored in one variable) looks fragile. It's correct, through. A simple if-then-else would, in my opinion, be better:
```C++
std::string popupmsg;
if( amount < input_amount ) {
    popupmsg = string_format( "...", count );
} else {
    popupmsg = string_format( "...", amount, count );
}
```
It's fragile because the format specifier in the string the actual parameters to `string_format` are separated. If the condition for how to chose the string is changed, but the condition below on how to call `string_format` is not changed accordingly, those might not match anymore and cause problems with format specifiers not matching the actual parameters.